File: Wire-iOS/Sources/Components/Buttons/Button.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 



#import "Button.h"

// helpers

#import "UIImage+ImageUtilities.h"
#import "Wire-Swift.h"

@import QuartzCore;

@implementation Button

+ (instancetype)buttonWithStyle:(ButtonStyle)style
{
    return [[Button alloc] initWithStyle:style];
}

- (instancetype)initWithStyle:(ButtonStyle)style
{
    return [self initWithStyle:style variant:ColorScheme.defaultColorScheme.variant];
}


- (instancetype)init
{
    self = [super init];
    
    if (self) {
        _borderColorByState = [NSMutableDictionary dictionary];
        _originalTitles = [[NSMutableDictionary alloc] init];
        self.clipsToBounds = YES;
    }
    
    return self;
}

- (CGSize)intrinsicContentSize
{
    CGSize s = [super intrinsicContentSize];
    
    return CGSizeMake(s.width + self.titleEdgeInsets.left + self.titleEdgeInsets.right,
                      s.height + self.titleEdgeInsets.top + self.titleEdgeInsets.bottom);
}

- (void)setBounds:(CGRect)bounds
{
    [super setBounds:bounds];
    
    [self updateCornerRadius];
}

- (void)setBackgroundImageColor:(UIColor *)color forState:(UIControlState)state
{
    [self setBackgroundImage:[UIImage singlePixelImageWithColor:color] forState:state];
}

- (UIColor *)borderColorForState:(UIControlState)state
{
    UIColor *borderColor = self.self.borderColorByState[@(state)];
    
    if (borderColor == nil) {
        borderColor = self.borderColorByState[@(UIControlStateNormal)];
    }
    
    return borderColor;
}

- (void)updateBorderColor
{
    self.layer.borderColor = [self borderColorForState:self.state].CGColor;
}

- (void)setTextTransform:(TextTransform)textTransform
{
    _textTransform = textTransform;
    
    [self.originalTitles enumerateKeysAndObjectsUsingBlock:^(NSNumber *state, NSString *title, BOOL *stop) {
        [self setTitle:title forState:state.unsignedIntegerValue];
    }];
}

- (void)setCircular:(BOOL)circular
{
    _circular = circular;
    
    if (circular) {
        self.layer.masksToBounds = YES;
        [self updateCornerRadius];
    } else {
        self.layer.masksToBounds = NO;
        self.layer.cornerRadius = 0;
    }
}

- (void)updateCornerRadius
{
    if (self.circular) {
        self.layer.cornerRadius = self.bounds.size.height / 2;
    }
}

#pragma mark - Observing state

- (void)setHighlighted:(BOOL)highlighted
{
    UIControlState previousState = self.state;
    [super setHighlighted:highlighted];
    [self updateAppearanceWithPreviousState:previousState];
}

- (void)setSelected:(BOOL)selected
{
    UIControlState previousState = self.state;
    [super setSelected:selected];
    [self updateAppearanceWithPreviousState:previousState];
}

- (void)setEnabled:(BOOL)enabled
{
    UIControlState previousState = self.state;
    [super setEnabled:enabled];
    [self updateAppearanceWithPreviousState:previousState];
}

- (void)updateAppearanceWithPreviousState:(UIControlState)previousState
{
    if (self.state == previousState) {
        return;
    }
    
    // Update for new state (selected, highlighted, disabled) here if needed
    [self updateBorderColor];
}


@end

File: Wire-iOS/Sources/Components/Buttons/Button.swift
//
// Wire
// Copyright (C) 2019 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import Foundation

extension Button {
    @objc
    convenience init(style: ButtonStyle, variant: ColorSchemeVariant) {
        self.init()
        
        textTransform = .upper
        titleLabel?.font = .smallLightFont
        layer.cornerRadius = 4
        contentEdgeInsets = UIEdgeInsets(top: 4, left: 16, bottom: 4, right: 16)
        
        switch style {
        case .full:
            setBackgroundImageColor(.accent(), for: .normal)
            setTitleColor(UIColor.white, for: .normal)
            setTitleColor(UIColor.from(scheme: .textDimmed, variant: variant), for: .highlighted)
        case .fullMonochrome:
            setBackgroundImageColor(UIColor.white, for: .normal)
            setTitleColor(UIColor.from(scheme: .textForeground, variant: .light), for: .normal)
            setTitleColor(UIColor.from(scheme: .textDimmed, variant: .light), for: .highlighted)
        case .empty:
            layer.borderWidth = 1
            setTitleColor(UIColor.buttonEmptyText(variant: variant), for: .normal)
            setTitleColor(UIColor.from(scheme: .textDimmed, variant: variant), for: .highlighted)
            setTitleColor(UIColor.from(scheme: .textDimmed, variant: variant), for: .disabled)
            setBorderColor(UIColor.accent(), for: .normal)
            setBorderColor(UIColor.accentDarken, for: .highlighted)
            setBorderColor(UIColor.from(scheme: .textDimmed, variant: variant), for: .disabled)
        case .emptyMonochrome:
            setBackgroundImageColor(UIColor.clear, for: .normal)
            setTitleColor(UIColor.white, for: .normal)
            setTitleColor(UIColor.from(scheme: .textDimmed, variant: .light), for: .highlighted)
            setBorderColor(UIColor(white: 1.0, alpha: 0.32), for: .normal)
            setBorderColor(UIColor(white: 1.0, alpha: 0.16), for: .highlighted)
        default:
            break
        }
    }

    open override func setTitle(_ title: String?, for state: UIControl.State) {
        var title = title
        state.expanded.forEach(){ expandedState in
            if title != nil {
                originalTitles?[NSNumber(value: expandedState.rawValue)] = title
            } else {
                originalTitles?.removeObject(forKey: NSNumber(value: expandedState.rawValue))
            }
        }
        
        if textTransform != .none {
            title = title?.applying(transform: textTransform)
        }
        
        super.setTitle(title, for: state)
    }
    
    @objc(setBorderColor:forState:)
    func setBorderColor(_ color: UIColor?, for state: UIControl.State) {
        state.expanded.forEach(){ expandedState in
            if color != nil {
                borderColorByState[NSNumber(value: expandedState.rawValue)] = color
            }
        }
        
        updateBorderColor()
    }
}
